### PR TITLE
Game status

### PR DESCRIPTION
**Pull request by Middlebrooks314**
Game Status ticket - Is the game over? -> Is there a winning combination or a full board?

https://trello.com/c/ChNj6d83

### CODE REVIEW

File: lib/board.rb
Comment (aaizuss) [inline lib/board.rb]: Should the board know about how to win tic tac toe? (I can go either way on this)

File: lib/board.rb
Comment (aaizuss) [inline lib/board.rb]: I asked about this in the previous pr, but do you think it's redundant to use the `board` suffix on methods defined in the `board` class? When these are used in your program, you'll see things like `board.mark_board`, `board.full_board?`, etc. Do you prefer that to `board.mark`, and `board.full?`

File: spec/board_spec.rb
Comment (aaizuss) [inline spec/board_spec.rb]: Since all you care about in this test is starting with a full board, you could write a loop that calls `mark_board` for every index and make the test shorter. Of course this is fine too

File: spec/board_spec.rb
Comment (aaizuss) [inline spec/board_spec.rb]: I wonder about mentioning `array` in the test description because it is so specific to implementation details, and we want our tests to focus on behavior. You could instead write `returns true when every position on the board is has a mark`. This is nitpicky but I'm curious about what others think.

File: spec/board_spec.rb
Comment (aaizuss) [inline spec/board_spec.rb]: You wrote 2 tests for winning combinations (diagonal and vertical) - are you confident that your method works for all winning combinations? Also, it would be helpful if your `it` blocks said what kind of winning combination it was (instead of distinguishing the combinations by the winning mark)

File: spec/game_spec.rb
Comment (aaizuss) [inline spec/game_spec.rb]: Does a board win? Do you mean "returns true if the board has a winner"?

File: spec/game_spec.rb
Comment (aaizuss) [inline spec/game_spec.rb]: I wouldn't expect to need to create a fake console in order to test the `game_over` method... what does this tell you about the responsibility of `Game`/the way your classes are set up?

**Review by aaizuss — COMMENTED** [review]
There are some opportunities for refactoring here, but it might be easier for you to refactor once you start working on your next story. I'm going to wait to approve until you answer some of my questions.

File: lib/board.rb
Comment (Middlebrooks314) [reply lib/board.rb]: I was conflicted about this, but I made that decision because I felt it was the board's job to know what was on the board and where the pieces were.

File: lib/board.rb
Comment (Middlebrooks314) [reply lib/board.rb]: I'm trying to be super explanatory with the method names, but I guess that it should be assumed that if you're in the board class that it's about the board.

File: spec/board_spec.rb
Comment (Middlebrooks314) [reply spec/board_spec.rb]: I was conflicted about this as well because I felt like I couldn't say "when the board is full" because what does that even mean programmatically? Also, it was redundant with the method name? I think the suggestion is valid, I can also wait for another reviewer's feedback.

File: spec/board_spec.rb
Comment (Middlebrooks314) [reply spec/board_spec.rb]: What would your suggestion be for the number of combinations? I can be more descriptive in the ` it` block

File: spec/game_spec.rb
Comment (Middlebrooks314) [reply spec/game_spec.rb]: What about a winning combination? I didn't write "a winner" because what is a winner? I figured winner implied a player. (nitpicky answer :upside_down_face:)

File: spec/game_spec.rb
Comment (Middlebrooks314) [reply spec/game_spec.rb]: Would you suggest that when my Game class instantiates that it can have a default argument?

File: spec/board_spec.rb
Comment (aaizuss) [reply spec/board_spec.rb]: Maybe you should say, returns true when there is a vertical winner, returns true when there is a diagonal winner, returns true when there is a horizontal winner - that kind of thing. And I would have more confidence in the test if you had a board set up where you have 2 Xs in a row somewhere with an O in between, and an O winning somewhere on the board.

File: spec/game_spec.rb
Comment (aaizuss) [reply spec/game_spec.rb]: It's up to you. I'm just trying to get you to think more critically about what your tests are communicating and how your code reads

File: spec/game_spec.rb
Comment (aaizuss) [reply spec/game_spec.rb]: No, that wouldn't be helpful here because you'd still have to mock out the console. I'm suggesting that you might need to make another class for something because it doesn't make sense that the `game_over` method implicitly depends on having a console.

File: lib/rules.rb
Comment (cmvandrevala) [inline lib/rules.rb:3]: Out of curiosity, why is Rules a module and not a class? Not saying this is right or wrong, just curious.

File: spec/board_spec.rb
Comment (cmvandrevala) [inline spec/board_spec.rb]: You don't need multiple `describe 'full?' do` lines. You can put these three tests under one `describe 'full?' do` block.

File: spec/rules_spec.rb
Comment (cmvandrevala) [inline spec/rules_spec.rb]: Same here. No need for multiple describe blocks - one is plenty

File: lib/rules.rb
Comment (cmvandrevala) [inline lib/rules.rb]: Not 100% sure if this will work, but Ruby might let you do something cool:

```
    WINNING_COMBOS.any? do |x,y,z|
      board.position_taken?(x) &&
      board.squares[x] == board.squares[y] &&
      board.squares[x] == board.squares[z]
    end
```

**Review by cmvandrevala — CHANGES_REQUESTED** [review]

File: lib/rules.rb
Comment (Middlebrooks314) [reply lib/rules.rb:3]: I chose to use it as a sort of static class since it is a small collection of rules methods. I didnt see at this point the Rules module having to keep track of any state because that is the job of the Game class and the Board class.

File: spec/board_spec.rb
Comment (aaizuss) [inline spec/board_spec.rb]: 🔍 typo in squares

File: spec/board_spec.rb
Comment (aaizuss) [reply spec/board_spec.rb]: Yeah - the convention is that you describe the method you're testing, and then use it for different expectations. You could also use context if you want to differentiate between different behaviors within a describe block

File: spec/board_spec.rb
Comment (aaizuss) [inline spec/board_spec.rb]: You have an extra word in each of your it blocks ("is has an X mark")

**Review by aaizuss — APPROVED** [review]

**Review by cmvandrevala — APPROVED** [review]

**Pull request merged**
